static/app/components/events/interfaces/sourceMapsDebuggerModal.tsx

```diff
@@ -0,0 +1,1033 @@
+
+  if (facts.stackFrameDebugId === null) {
+    return <CheckListItem status="none" title={itemName} />;
+  }
+
+  if (facts.uploadedSourceFileWithCorrectDebugId) {
+    return <CheckListItem status="checked" title={itemName} />;
+  }
+
+  if (facts.uploadedSomeArtifactWithDebugId) {
+    return (
+      <CheckListItem status="alert" title={itemName}>
+        <CheckListInstruction type="muted">
+          <h6>No Soure File With Matching Debug ID</h6>
+          <p>
+            You already uploaded artifacts with Debug IDs but none of the uploaded source
+            files had a Debug ID matching this stack frame's Debug ID:{' '}
+            {<MonoBlock>Debug ID: {facts.stackFrameDebugId}</MonoBlock>}
+          </p>
+          <p>
+            Make sure to inject Debug IDs into all of your source files and to upload all
+            of them to Sentry.
+          </p>
+          {/* TODO: Link to Uploaded Artifacts */}
+        </CheckListInstruction>
+      </CheckListItem>
+    );
+  }
+
+  if (facts.uploadedSomeArtifact) {
+    return (
+      <CheckListItem status="alert" title={itemName}>
+        <CheckListInstruction type="muted">
+          <h6>Uploaded Artifacts Without Debug IDs</h6>
+          <p>
+            You already uploaded artifacts for this project but none of the artifacts
+            contain Debug IDs. Make sure you inject Debug IDs into your source files
+            before uploading them to Sentry.
+          </p>
+          <p>
+            Read the{' '}
+            <ExternalLink href="https://docs.sentry.io/platforms/javascript/sourcemaps/">
+              Sentry Source Maps Documentation <IconOpen size="xs" />
+            </ExternalLink>{' '}
+            to learn how to inject Debug IDs into your build artifacts and how to upload
+            them to Sentry.
+          </p>
+          {/* TODO: Link to Uploaded Artifacts */}
+        </CheckListInstruction>
+      </CheckListItem>
+    );
+  }
+
+  return (
+    <CheckListItem status="alert" title={itemName}>
+      <CheckListInstruction type="muted">
+        <h6>No Artifacts Uploaded</h6>
+        <p>
+          You didn't upload any artifacts yet. Read the{' '}
+          <ExternalLink href="https://docs.sentry.io/platforms/javascript/sourcemaps/">
+            Sentry Source Maps Documentation <IconOpen size="xs" />
+          </ExternalLink>{' '}
+          to learn how to inject Debug IDs into your build artifacts and how to upload
+          them to Sentry.
+        </p>
+        {/* TODO: Link to Uploaded Artifacts */}
+      </CheckListInstruction>
+    </CheckListItem>
+  );
+}
+
+function UploadedSourceMapWithCorrectDebugIdChecklistItem({
+  facts,
+}: {
+  facts: SourceResoultionResults;
+}) {
+  const itemName = 'Uploaded source map with a matching Debug ID';
+
+  if (facts.stackFrameDebugId === null || !facts.uploadedSourceFileWithCorrectDebugId) {
+    return <CheckListItem status="none" title={itemName} />;
+  }
```

Comment:
imho the single checklist items should not be aware of how to measure the progress of it's successor. This makes it harder to maintain.
Could we determine the state of the single items by checking the progress count?
```ts
function getItemStatus(progress: number, itemIndex: number): 'checked' | 'alert' | 'none' {
  if (itemIndex < progress) {
    return 'checked'
  }
  if (itemIndex === progress) {
    return 'alert'
  }
  return 'none'
}
```

Then the first check in the single items would become a lot simpler:
```ts
if (!status === 'alert') {
  return <CheckListItem status={status} title={itemName} />;
}
```

All this under the assumption that only one item can be the active one and that it is not possible to skip steps.

Reply:
I agree with you. I hoisted the state into the parent component. Hopefully this makes everything a bit more maintainable. [bfe7747](https://github.com/getsentry/sentry/pull/55659/commits/bfe77471519d5362c6d0e9d735b92802a8f2258c)